ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStore.java

```diff
@@ -65,15 +65,21 @@ public SlotStore(SlotStoreEnvironmentBean config) throws IOException {
 
         // unused for now, but eventually we'll have a disk backend...
         File storeDir = new File(config.getStoreDir());
-        if (!storeDir.exists() && !storeDir.mkdirs()) {
-            throw new IOException(tsLogger.i18NLogger.get_dir_create_failed(storeDir.getCanonicalPath()));
-        }
         storeDirCanonicalPath = storeDir.getCanonicalPath();
 
-        slots = new RAMSlots(config);
+        slots = config.getBackingSlots();
+        slots.init(config);
 
+        // internal recovery to rebuild the slotIdIndex and freeList
         for (int i = 0; i < config.getNumberOfSlots(); i++) {
-            freeList.add(i); // TODO recovery, when the backend supports it.
+            byte[] data = slots.read(i);
+            if (data == null || data.length == 0) {
+                freeList.add(i); // slot does not contain a valid entry, is free for use
```

Comment:
Right. It's an implementation detail of the write. A partial write is never exposed, it just behaves as never having happened, so the slot is considered to still be empty. There is nothing useful a user can do about it even if it's made visible to them, so they don't need to know. The mechanism that puts overwrites for the same key into parallel slots is critical here, as if we let data vanish by partially overwriting it then crashing, that would be a user visable data loss that would not be recoverable.

Reply:
> Right. It's an implementation detail of the write. A partial write is never exposed, it just behaves as never having happened, so the slot is considered to still be empty. 
 
See my earlier comment about a full write, a read followed by a partial rewrite. That sequence has exposed the data to the outside world.

> There is nothing useful a user can do about it even if it's made visible to them, so they don't need to know. 
 
A log message may enable them to correlate it with earlier log messages.

> The mechanism that puts overwrites for the same key into parallel slots is critical here, as if we let data vanish by partially overwriting it then crashing, that would be a user visable data loss that would not be recoverable.

So are you saying that a slot update that goes into the parallel new slot can never become corrupt after it has been written. If this is an improbable scenario then your counter argument is valid. But should our transaction system strive to deal with improbable scenarios or would you argue that dealing with this particular scenario (of corruption after the fact) would be overkill. Either way a log message to say that a corrupt slot was encountered would deal with the circumstance.

Reply:
> full write, a read followed by a partial rewrite. That sequence has exposed the data to the outside world.

It has exposed the full write to the read, which is fine. The partial write is never exposed. A read before will read the previous full state. A read after will likewise read the previous full state, which still exists because it's in a different slot. The state is logically atomically overwritten, but physically never partially overwritten - internally the old and new states exist in parallel.

Reply:
Yes but my point is that the slot was read and real world decisions may have been made, for example XA resources have been contacted. When recovery later tries to read the slot and it is corrupted the implementation quietly ignores it and the resources may be left in limbo.

Reply:
So this conversation was resolved by a [later comment](https://github.com/jbosstm/narayana/pull/1593#issuecomment-614521862).